protocols/smart-transfer-protocol.asyncapi.yaml

```diff
@@ -0,0 +1,424 @@
+
+  dataset/pull/metadata/response:
+    publish:
+      operationId: odf.dataset.pull.metadata.response
+      summary: Response with the metadata for the requested pull range
+      security:
+        - jwt: []
+      message:
+        messageId: odf.dataset.DatasetMetadataPullResponse
+        name: DatasetMetadataPullResponse
+        contentType: application/json
+        payload:
+          type: object
+          required:
+            - blocks
+          properties:
+            blocks:
+              $ref: '#/components/schemas/ObjectsBatch'
+  
+  dataset/pull/objects/transfer/request:
+    subscribe:
+      operationId: odf.dataset.pull.objects.transfer.request
+      summary: Requests to agree on the strategy for downloading 1 or more object files
+      security:
+        - jwt: []
+      message:
+        messageId: odf.dataset.DatasetPullObjectsTransferRequest
+        name: DatasetPullObjectsTransferRequest
+        contentType: application/json
+        payload:
+          type: object
+          required:
+            - objectFiles
+          properties:  
+            objectFiles:
+              type: array
+              description: List of object file references in this bulk
+              items:
+                $ref: '#/components/schemas/ObjectFileReference'
+      
+  dataset/pull/objects/transfer/response:
+    publish:
+      operationId: odf.dataset.pull.objects.transfer.response
+      summary: Response on the agreed strategy for objects downloading during the pull
+      security:
+        - jwt: []
+      message:
+        messageId: odf.dataset.DatasetPullObjectsTransferResponse
+        name: DatasetPullObjectsTransferResponse
+        contentType: application/json
+        payload:
+          type: object
+          required:
+            - objectTransferStrategies
+          properties:   
+            objectTransferStrategies:
+              type: array
+              description: Agreed download operations for the requested bulk of object files
+              items:
+                $ref: '#/components/schemas/PullObjectTransferStrategy'
+
+  dataset/push/request:
+    subscribe:
+      operationId: odf.dataset.push.request
+      summary: Initiates dataset push to the ODF server
+      security:
+        - jwt: []
+      message:
+        messageId: odf.dataset.DatasetPushRequest
+        name: DatasetPushRequest
+        contentType: application/json
+        payload:
+          type: object
+          required:
+            - datasetId
+            - currentHead
+          properties:
+            datasetId:
+              type: string
+              format: did
+              description: Dataset identifier
+            currentHead:
+              type: string
+              format: multihash
+              description: Hash of the head block that client assumes to be the current head on the server
+            sizeEstimation:
+              # Question: should it stay optional?
```

Comment:
IMO we should start with required and then relax if needed. This data will be useful for rejecting pushes that put users above their storage/traffic quotas early.

I'd rather we fail request as early as possible than exhaust the user quotas completely with a transfer that was guaranteed to fail.

Reply:
Ok